src/java.base/share/classes/java/lang/Class.java

```diff
@@ -593,14 +598,80 @@ public boolean isInlineClass() {
      *         inline class; an empty {@link Optional} otherwise
      * @since Valhalla
      */
-    public Optional<Class<T>> referenceType() {
-        return valType != null ? Optional.ofNullable(refType) : Optional.of(this);
+    public Optional<Class<?>> referenceType() {
+        if (isPrimitive() || isInterface() || isArray())
+            return Optional.empty();
+        ensureProjectionTypesInited();
+        System.out.println(getName() + " " + Arrays.toString(projectionTypes));
+        return projectionTypes.length == 2 ? Optional.of(projectionTypes[1]) : Optional.empty();
     }
 
     // set by VM if this class is an inline type
     // otherwise, these two fields are null
     private transient Class<T> valType;
     private transient Class<T> refType;
+    private transient Class<?>[] projectionTypes;
+
+    private synchronized void ensureProjectionTypesInited() {
+        if (isPrimitive() || isArray() || isInterface())
+            return;
+
+        if (projectionTypes != null)
+            return;
+
+        if (isInlineClass()) {
+            Class<?> superClass = getSuperclass();
+            System.out.println("superclass " + superClass);
+            if (superClass != Object.class && superClass.isReferenceProjectionType()) {
+                projectionTypes = new Class<?>[] { this, superClass };
+            } else {
+                projectionTypes = new Class<?>[] { this };
+            }
+        } else if (isReferenceProjectionType()) {
+            projectionTypes = new Class<?>[] { valueProjectionType(), this };
```

Comment:
It seems odd that isReferenceProjectType() uses projectionTypes before it is initialized (or is it re-initialized).
(Though there is a check for null).

Reply:
I have redone the logic and incorporated your suggestion in `getProjectionType` method.